LinkClock quantum handling and meter sync

jamshark70 opened the pull request:
## Purpose and Motivation

See #4401.

1. We should not set Link's quantum when setting the SC clock's `beatsPerBar`. This PR adds `quantum` and `quantum_` so that the user can control it directly (and *not* change it at an inappropriate time).
   - A valid use of `quantum_` is: If you know, before performing a piece, that it will be in 3/4 time throughout, you can do `clock.quantum = 3` on all SC LinkClocks, and Link will synchronize the barlines.
   - *But*, if you change meter during the performance, you can't change `quantum` (or beats will be discontinuous), and the discrepancy between quantum and beatsPerBar means that LinkClocks to join later may or may not be barline-synced.
2. To handle discrepancies, users may now call `syncMeter_(true)` on a LinkClock and it will do two things:
   a. Look for other SC LinkClocks on the network that are also set to sync meter, and align the local `beatsPerBar` and `baseBarBeat` to match. Currently playing tasks will keep going as they were (there is no disruption in beats or beat durations), but future scheduling relative to bars (e.g. quant = -1) will be in sync with the other SC peers.
   b. Future meter changes will be broadcast to other syncMeter peers.

A use case for syncMeter is: You have a performance that requires a meter change midway through (so, at some point, beatsPerBar and quantum will be different). Suppose then that an SC client fails during the performance and has to rejoin the network. Without syncMeter, it isn't guaranteed to be aligned with the other SC peers -- but, if that machine rejoins with `TempoClock.default = LinkClock.new.syncMeter_(true)` then it's all taken care of automatically.

## Types of changes

- Documentation
- New feature
- Breaking change ("technically" a breaking change in that, previously, `setMeterAtBeat` would set Link's quantum, and this PR removes that. But it was a mistake before and users shouldn't be relying on it.)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested (PR includes unit tests)
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

jamshark70 commented:
I had tested previously with multiple sclang instances on one machine. Now tested with two machines, working well.

geoffroymontel commented:
Dear James. I've just tried this branch to sync Ableton Live with SC and it's working great ! Thanks for this really valuable addition !

jamshark70 commented:
@geoffroymontel thanks for testing! Glad to hear that it's working.

dataexcess commented:
Hi, 

Just wondering why this awesome feature is not yet merged to the master branch? I am now creating all my projects on this branch, but therefor missing out on all the latest features and bufixes :(...

brianlheim commented:
hi @dataexcess, the answer is pretty simple -- we have a backlog of PRs waiting to be reviewed and merged right now. Someone will get to this one eventually!

In the meanwhile, you can get the other features and bugfixes for yourself by rebasing this branch on the latest develop and rebuilding.

dataexcess commented:
> hi @dataexcess, the answer is pretty simple -- we have a backlog of PRs waiting to be reviewed and merged right now. Someone will get to this one eventually!
> 
> In the meanwhile, you can get the other features and bugfixes for yourself by rebasing this branch on the latest develop and rebuilding.

Ahh of course! thnx for the tip.

jamshark70 commented:
> I am now creating all my projects on this branch...

Out of curiosity, which feature(s) are you using? Asking because this PR is relevant only when changing beatsPerBar on the fly.

If you're always in 4/4, then Link should sync the barlines without an extra `syncMeter`.

If you're always in 3/4, I believe in the current release, you can set beatsPerBar to 3 in all peers before starting the performance -- currently this sets the quantum as well, which may break some beat durations but that's ok if you haven't started playing yet.

If you are changing beatsPerBar midstream, then you need this branch. I'm pretty sure, if you're not doing that, you don't really need the new stuff here.

dataexcess commented:
> > I am now creating all my projects on this branch...
> 
> Out of curiosity, which feature(s) are you using? Asking because this PR is relevant only when changing beatsPerBar on the fly.
> 
> If you're always in 4/4, then Link should sync the barlines without an extra `syncMeter`.
> 
> If you're always in 3/4, I believe in the current release, you can set beatsPerBar to 3 in all peers before starting the performance -- currently this sets the quantum as well, which may break some beat durations but that's ok if you haven't started playing yet.
> 
> If you are changing beatsPerBar midstream, then you need this branch. I'm pretty sure, if you're not doing that, you don't really need the new stuff here.

My bad! LinkClock is included in the bleeding edge build, and I was running the '3.10.3 - apple notarization signed build'. 🤦‍♂️ I wasn't really worried about these features.. I thought this was the main LinkClock branch. Nonetheless, changing beatsPerBar midstream would be cool to be able to do :)

brianlheim commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+				// So we have to do a two-step check.
+				if(bpbs.size == 1) {
+					// 'choose' = easy way to get one item from an unordered collection
+					newBeatsPerBar = bpbs.choose;
```
`pop` is better, i think. `choose` randomly probes the backing array until a valid item is found, while `pop` will linearly scan. since items are inserted linearly into the backing array, `pop` will return after checking the first index while `choose` may not

jamshark70 replied:
Good catch, will do.

brianlheim commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+						// myBeats = replies.choose[\queriedAtBeat].round(round);
+						// theirPhase = bibs.choose;  // should be only one
+						newBase = baseBeats.choose;
+						if(verbose) { "syncing meter to %, base = %\n".postf(bpbs.choose, newBase) };
```
this could be `newBeatsPerBar`

brianlheim replied:
i'm generally skeptical of printing informative text by default. can you talk a little bit about this one?

jamshark70 replied:
> this could be newBeatsPerBar

Good point -- and now that it's `.pop` (which modifies the set), it *has* to be.

> i'm generally skeptical of printing informative text by default.

There is one concrete reason to post. Suppose you already have one SC LinkClock with a meter sync object with beatsPerBar = 3, and on another sclang instance, you start another LinkClock, with meter sync, and default settings (beatsPerBar = 4). Meter sync demands that the new clock join the 3/4 meter (and, if the network is functioning, it will), but that isn't what the code explicitly requested. So the user may have thought she was asking for normal 4/4 time, but ended up with 3/4 (because meter sync means "your requested meter, or others' meter if available"). *Maybe* it's better to do that silently, but I think it's better to advise the user.

I could change this to `if(verbose and: { newBeatsPerBar != clock.beatsPerBar })` perhaps.

brianlheim commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+						// calculate baseBarBeat such that my local beatInBar will match theirs
+						// myBeats = replies.choose[\queriedAtBeat].round(round);
+						// theirPhase = bibs.choose;  // should be only one
+						newBase = baseBeats.choose;
```
again pop. also, the comment above doesn't really make sense to me

brianlheim commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+						this.prSetMeterAtBeat(bpbs.choose, newBase);  // local only
+					} {
+						// this should not happen
+						Error("Discrepancy among 'syncMeter' Link peers; cannot sync barlines").throw;
```
perhaps the error message should be different to distinguish it in case and severity from the one below?

brianlheim commented on lang/LangPrimSource/SC_LinkClock.cpp:
```diff
@@ -113,6 +113,19 @@ int prLinkClock_NumPeers(struct VMGlobals* g, int numArgsPushed) {
+    if (!clock) {
+        error("clock is not running.\n");
+        return errFailed;
+    }
```
this matches surrounding code but would prefer to see some of this factored out if you have time, the same 5 lines are repeated in every primitive here

jamshark70 replied:
I don't mind to change it, but how should it be refactored? If there's a `getClockInstance()` function, then you could pass in either `g->sp` or `g->sp - 1`, easy enough, but then... if you return this, then the calling function still needs to have an `if` block to post the error and return. Or, leave the stack access as is and the factored-out function could handle the null pointer case -- but the calling function would still need an `if` block to return the error code to its caller.

Sure, code duplication is usually a red flag, but at my admittedly low C++ skill level, I don't see how factoring it out will make it more concise (and if it doesn't make it more concise, then what's the point?).

brianlheim replied:
one simple idea -

```cpp
LinkClock* tryGetLinkClock(PyrSlot* clockSlot) {
    auto* clock = static_cast<LinkClock*>(slotRawPtr(slotRawObject(clockSlot)->slots + 1));
    if (!clock) {
        error("clock is not running.\n");
    }
    return clock;
```

```cpp
int prLinkClock_GetQuantum(...) {
    if (auto* clock = tryGetLinkClock(g->sp)) {
        SetFloat(g->sp, clock->GetQuantum());
        return errNone;
    } else {
        return errFailed;
    }
}
```

---

another idea, kind of like a continuation, doesn't contain an extra branch like the idea above. if we had std::optional with a monadic interface we could use `map()`

```cpp
// Expected signature for F: `int(LinkClock*)`, returning an error code
template <typename F>
void getLinkClockAndThen(PyrSlot* clockSlot, F&& continuation) {
    auto* clock = static_cast<LinkClock*>(slotRawPtr(slotRawObject(clockSlot)->slots + 1));
    if (clock) {
        return continuation(clock);
    } else {
        error("clock is not running.\n");
        return errFailed;
    }
}
```

```cpp
int prLinkClock_GetQuantum(...) {
    getLinkClockAndThen(g->sp, [](const LinkClock* clock) {
        SetFloat(g->sp, clock->GetQuantum());
        return errNone;
    });
}
```

```cpp
int prLinkClock_SetQuantum(...) {
    getLinkClockAndThen(g->sp - 1, [g](LinkClock* clock) {
        double quantum;
        if (slotDoubleVal(g->sp, &quantum)) {
            return errWrongType;
        } else {
            clock->SetQuantum(quantum);
            return errNone;
        }
    });
}
```

---

you could even get it down to something like

```cpp
int prLinkClock_GetQuantum(...) { return tryGetLinkClockField(g->sp, &LinkClock::GetQuantum); }
int prLinkClock_SetQuantum(...) { return trySetLinkClockField(g->sp, &LinkClock::SetQuantum); }
int prLinkClock_GetLatency(...) { return tryGetLinkClockField(g->sp, &LinkClock::GetLatency); }
int prLinkClock_SetLatency(...) { return trySetLinkClockField(g->sp, &LinkClock::SetLatency); }

// at which point you may as well refactor it so you can write this:
definePrimitive(..., "_LinkClock_GetQuantum", &tryGetLinkClockField<LinkClock::GetQuantum>);
```

but that would depend on type assumptions that are a bit fragile.

happy to explain any of this further if you'd like me to. also feel free to not do any of it in this PR.

jamshark70 replied:
I ended up taking the first, simplest option. I agree that it reads better.

I'm not interested enough in C++ to wrap my head around that last one :rofl: The continuation does make sense to me (since I'm accustomed to first class functions in sclang) but I lean towards a simpler C++ idiom. (One reason, for instance, why I haven't touched supernova code is that I just don't understand it, and don't have time or interest to get there.)

brianlheim replied:
> I'm not interested enough in C++ to wrap my head around that last one 🤣

tbh, i find it a little frustrating that you asked me to help you understand how to refactor this, and are now saying you're not interested in understanding it after i made an attempt to explain it.

> I lean towards a simpler C++ idiom. 

so do i, which is why i prefer the solution that uses standard library features to eliminate a large amount of duplicated code.

jamshark70 replied:
Oh, now I see the "reply" field... it wasn't there before.

I apologize for causing offense -- I had read the last couple of suggestions like "well, you *could* do this but *haha* of course we won't," and my intent was to respond in a similarly lighthearted way, but I see how my wording could be read differently. (I might also suggest that PR review comments could ideally stick closer to what is required to resolve the PR, without an expectation that anything extra would be taken with the same weight. The only parts that I said I wasn't interested in were rather exotic suggestions that we would never have used.)

I incorporated the first of those suggestions -- agreed that it reads better this way.

jamshark70 commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
 	}
 
+	init { |tempo, beats, seconds, queueSize|
+		super.init(tempo, beats, seconds, queueSize);
```
> are users expected to call this themselves?

No, certainly not.

It's a general convention throughout the class library that a constructor method calls `init`, and users don't do it manually. That's dating back to JMc's original class library implementation, and in the 17 years I've been involved in SC, I can't remember even one case where user got into trouble by thinking "oh gosh, there's an init method, I had better call it."

So IMO, non-issue.

brianlheim replied:
then my question is how does this get called? history lesson aside, i don't see it being called anywhere.

jamshark70 replied:
It gets called by the constructor.

Method selectors always resolve against the class itself or the nearest superclass. If a class inherits *new from a superclass, and *new does the standard `super.new.init`, and the class implements an instance method `init`, then *new gets answered by the superclass implementation while the subclass handles init. It doesn't matter that the *new implementation is in the superclass. The only thing that is known at the time of calling init is that the receiver is an instance of the specific class and that it implements init, and it's this implementation that takes control.

To do the superclass's init, it's necessary to forward up to `super.init` (as is done here).

This is all standard for SC.

brianlheim replied:
thanks for the explanation

brianlheim replied:
for future reference, the concepts in your last comment are all ones i am familiar with, and i do not need them re-explained. i didn't do a good job of wording it, but my question was really just "who is calling init"/"are you sure there's something calling it". i gather now from your answer and from looking at TempoClock that init is indeed being invoked.

jamshark70 replied:
Apologies for misunderstanding the question. I thought the question was either "why is it not being called from within this code file?" or "if the constructor is in the superclass, isn't it calling superclass init and not using this one?" and I was just trying to head off foreseeable follow-up questions.

jamshark70 commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
 
+	init { |tempo, beats, seconds, queueSize|
+		super.init(tempo, beats, seconds, queueSize);
+		id = 0x7FFFFFFF.rand;  // to ignore 'self' queries
```
That's a good point (though I'd suggest, in a typical use case, that users should initialize clocks across the network before pulling random numbers for musical purposes).

I can also imagine an ensemble may just assign each member an id that they use every time, so, making it configurable is a good idea.

Between configuring this and network ports (which aren't a part of Link's protocol, nor of normal clock operation), I'm thinking it would be better to split meter sync into a separate class (to avoid god-object design). So instead of `LinkClock.new.syncMeter_(true)`, you might do:

```
l = LinkClock.new etc;
m = SCLinkClockMeterWatcher(l);
```

... and address commands to the separate sync object. I think it makes more sense than cramming tangentially related options into LinkClock itself.

brianlheim replied:
> I'm thinking it would be better to split meter sync into a separate class (to avoid god-object design)

yes, that's a great idea! i hadn't even considered that

jamshark70 replied:
Ok, I'll try to do that next week, and incorporate the other recommendations at that time. (I generally agree with the other review comments, won't reply to them all individually but I'm noting them to include in v2.)

brianlheim replied:
sounds good! fyi, when there are a lot of comments on a PR i find using the :+1: reaction and/or "resolve conversation" buttons helpful to keep track of what has and hasn't been dealt with yet.

brianlheim replied:
(sorry if you already do that, i've seen some people not do it and with PRs this big it can generate quite the sprawl)

jamshark70 commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -11,6 +13,169 @@ LinkClock : TempoClock {
+		meterQueryResp = OSCFunc({ |msg|
+			NetAddr.broadcastFlag = true;
+			// maybe some peers had to take a different port
+			(57120 .. 57127).do { |port|
```
Agreed, see other comment.

jamshark70 commented on SCClassLibrary/External/Ableton/LinkClock.sc:
```diff
@@ -1,6 +1,8 @@
 
 LinkClock : TempoClock {
+	var <syncMeter = false, <id,
+	meterChangeResp, meterQueryResp;
```
Finally working through these.

> also, why is id publicly accessible?

Now it's a deliberate decision -- if `id` is to be user-configurable in the new class, then it should also be accessible. (Your instinct to ask about it was right, though -- I had exposed it mainly for debugging during development, so there wasn't a good reason. Now there is.)

jamshark70 commented:
Just pushed most of the review fixes. Still remaining to refactor resyncMeter.

I'm writing documentation but it's slow going.

Unit tests need to be revised.

jamshark70 commented:
I believe I have addressed all review comments now. Documentation and unit tests are updated.

Oh, one comment is pending (regarding the repeated code to access the clock-object pointers in the primitives). Anything I can think of for that is not an improvement. Suggestions are welcome.

telephon commented on SCClassLibrary/External/Ableton/SCClockMeterSync.sc:
```diff
@@ -85,10 +85,16 @@ SCClockMeterSync {
-		resp;
+		var replies, resp;
 		if(clock.numPeers > 0) {
+			if(timeout.isKindOf(SimpleNumber).not or: { timeout <= 0 }) {
```
what would be passed instead of a number?

jamshark70 replied:
This is per a review comment from Brian: It's critical that this method not wait forever, but it's possible to pass in (erroneously) an object that will cause it to hang indefinitely. It's not a bad thing, I think, to prevent that before it happens.

telephon replied:
But wouldn't it be better to throw an error as usual?

jamshark70 replied:
Do we have an established practice of what to do if one thread is hanging for an asynchronous result, but the asynchronous operation fails with an exception?

I don't think we do. That will be the case here.

So, maybe I should add an `onError` argument to queryMeter.

telephon replied:
I think yes, or maybe just within `resyncMeter`, where it is used, have a protect:

```supercollider
protect {
	this.queryMeter { |val|
	replies = val;
	cond.unhang;
	}
} {
	cond.unhang;
}
```

jamshark70 replied:
The place where the error would occur is forked onto a separate thread. queryMeter has already returned, protect has already returned, you can't catch the error this way.

It's the same problem that I was trying to explain on the mailing list about Scott C.'s Promise-like class -- you can't throw an error directly back into the thread that was waiting. I ended up dealing with it by saving the error in a collection associated with the thread -- then, resuming the thread went back into the Promise's wait method, where we check for an error belonging to the thread and throw at that time.

But now we are getting out of scope of this PR, as we seem to love to do. This is actually kind of a tricky issue and I'd suggest that perhaps an already-large PR is not the place to establish a new best practice.

telephon replied:
>you can't throw an error directly back into the thread that was waiting.

I assumed that was the case, so no digression on my part. OK, we need to investigate into this. I thought that it does this (see `handleError`), but now have forgotten how I made it  work back then.

telephon replied:
> queryMeter has already returned
I don't see why it would return if an error is thrown in it.

But if you all think this is a bigger or different issue, forget about my suggestion.

jamshark70 replied:
> I don't see why it would return if an error is thrown in it.

The error is not thrown in the method!

It's thrown in a routine that happened to have been forked within the method.

Anywhere else in the codebase, if someone wrote `myMethod { fork { ... } }`, it would be understood that the sequence of events is:

1. The method schedules a routine on a clock.

2. The method returns `this`.

3. When the interpreter is not busy, the scheduled routine awakes.

The error will be thrown in stage 3. *Not* 1 or 2.

It's too late to bounce the error back to the caller because the caller is already forgotten (stack frames already traced back).

jamshark70 replied:
My suggestion here is to mark the location in the codebase with some comment that we can grep later, and open an issue for the broader question of catching errors thrown in forked routines (i.e., we are in a forked routine, where a Condition or such is `hang`ing, and the thing it's waiting for needs to throw an error back), then circle back and update here.

I can imagine how to do it by adding something to Condition, but then this PR would become responsible not only for a new class, documentation and unit tests directly related to the issue at hand, but also for new methods, documentation and unit tests for something not really related... let alone, as Brian said on the mailing list, sync issues are notoriously difficult to get right. I think in this case (perhaps not every case), the design considerations should not be tacked onto an unrelated changeset.

telephon replied:
yes, but there are details. let's keep the details for another discussion!

jamshark70 replied:
Oh, wait, I see what you mean... I was thinking of Brian's earlier comment that the "wait" expression would throw (which is inside the fork) but we could just throw from this `if`, before forking anything. (After literally 6-7 hours of addressing review comments, my mind was apparently a bit jelly.)

It probably is better to throw from here. Will do next time I'm at the computer.

telephon replied:
yes, that was actually what I meant. Thank you for taking a look!

jamshark70 replied:
Done.

jamshark70 commented:
I think that's everything. Passing tests.

I added one more feature: `repeats` and `delta` for messaging redundancy.

UDP is known to drop packets. I've seen this happen in an ensemble setting. So the problem is: If one machine broadcasts a meter change, and one of the machines doesn't get the message, then that machine will be out of sync. After that point, any `resyncMeter` attempts will fail, because there is a discrepancy among the peers.

That is, the logic in the class is sound, provided that communication always works. But we know that the communication doesn't always work (and a communication failure could be fatal).

So, with default settings now, every outgoing message is sent 4 times. (The user can configure this.) If there are machines A, B and C, then there are eight transmissions to consider, and to encounter the failure case, you would have to suppose that A --> B failed not once, but four times -- unlikely over a reasonably functioning network, *and* even if half of the transmissions failed (already unlikely), the odds of all of them being A --> B are slim.

(This means incoming OSC is redundant as well. I've handled that -- in normal cases, operations on the clock object should take place only once.)

I suppose a better way would be handshaking, but the problem there is that you have no way of knowing how many SuperCollider peers to expect to reply, or where they are (unlike TCP, where the connection is explicit). So I think it's a reasonable approach to throw more messages out, raising the probability of at least one of them getting through.

(Purely as an aside -- unit testing was very helpful here. When I added the messaging redundancy, at first I used `forkIfNeeded`, but this broke one of the unit tests. Because of that, I realized that `forkIfNeeded` meant that changing meter within a routine would cause `repeats * delta` beats to be added to the thread's timing -- wrong! And, fixed now. Just mentioning because, as much as I complain about unit tests, I do see their significance.)

brianlheim commented on SCClassLibrary/External/Ableton/SCClockMeterSync.sc:
```diff
@@ -0,0 +1,246 @@
+// but this may in theory apply to any clock that answers to
+// baseBarBeat_ (e.g. MIDISyncClocks, or BeaconClock)
+
+// TODO OSC message prefix with SC_
```
this appears to be done now

jamshark70 replied:
Ah, forgot to remove that. Will do in the next round.

brianlheim commented on SCClassLibrary/External/Ableton/SCClockMeterSync.sc:
```diff
@@ -0,0 +1,246 @@
+// TODO OSC message prefix with SC_
+
+SCClockMeterSync {
+	var <clock, <>id, <ports;
```
i don't think `id` should be settable once this object is constructed -- otherwise you leave open the possibility for it to be changed while messages are in flight. same with `ports`. it's probably simpler (i.e., less error prone) in those cases to just require users to construct a new object.

jamshark70 replied:
OK.

brianlheim commented on SCClassLibrary/External/Ableton/SCClockMeterSync.sc:
```diff
@@ -0,0 +1,246 @@
+	repeats_ { |value|
+		if(value.isKindOf(SimpleNumber).not or: { value < 1 }) {
+			Error("Invalid number of repeats '%'".format(value)).throw;
+		};
```
the value check i understand, but not the class validation.. isn't it more smalltalk-y to allow any object here that works like a number?

jamshark70 replied:
Ideally yes, but:

```
(
c = Complex(1, 0);

c.isNumber.debug("Complex is a number");

fork { c.wait };
)

Complex is a number: true
-> a Routine
ERROR: Message 'wait' not understood.
RECEIVER:
Instance of Complex
```

We're missing a polymorphic interface for `isWaitDuration`. I'm not quite sure how else to handle it in the context of this PR.

Incidentally SimpleNumber is the only number-like class that defines `wait`.

In the backend, threads are rescheduled if they yield something where `slotDoubleVal()` succeeds. As far as I can see, that eventually gets around to:

```
template <typename numeric_type> inline int slotVal(PyrSlot* slot, numeric_type* value) {
    if (IsFloat(slot)) {
        *value = static_cast<numeric_type>(slot->u.f);
        return errNone;
    } else if (IsInt(slot)) {
        *value = static_cast<numeric_type>(slot->u.i);
        return errNone;
    }
    return errWrongType;
}

... and...
inline bool IsFloat(const PyrSlot* slot) { return slot->tag == tagFloat; }
```

So we should allow only slots that are tagged float or int (i.e. SimpleNumber). Anything else will cause the thread to hang indefinitely. And, `anythingElse.wait` will throw a DoesNotUnderstand error, also causing the thread to hang indefinitely. So, this is ugly, but it might be the safest way to ensure the thread proceeds.

brianlheim replied:
that's fair. i guess my concern is, do we want to start auditing every bit of code in the class library for this? when i pointed out that the thread could end up erroring and hanging indefinitely, i wasn't thinking it meant it had to be avoided at all costs. i was noting it to see if you had any thoughts or opinions on it.

if you think this needs to be addressed this way, then i think we need to mitigate technical debt:
1. open a new issue describing the problem
2. agree on a workaround (what you have here seems fine)
3. comment everywhere the workaround is used so that when the issue opened in (1) is resolved, the workarounds can be easily spotted and replaced with the solution
4. in some order - compile a list of places where this safeguard is needed, and solve the underlying issue, and then safeguard unsafe `waits` relying on user-given values

jamshark70 replied:
I guess, if that's what is required, then we might as well just let it error out. That's more in keeping with SC's general attitude of allowing users to shoot themselves in the feet. (I did misunderstand your comment -- I thought you were asking for prevention.)

It makes me a little curious about garbage collection for hanging threads. In this case, if the wait time is valid, the queryMeter thread is reachable via the clock's queue array. The thread is running an open function whose containing scope has received a callback function from the resyncMeter thread, and that function refers to a Condition which contains the resyncMeter thread. I *suppose*, if the queryMeter thread timeout is not scheduled (by error or by inappropriate yield value), then the whole construct is unreachable and GC-able. If that's the case, then I'm fine with letting it die and then it's up to GC. But if we will have these stray actions hanging about forever, then we should more actively prevent it.

brianlheim replied:
> I guess, if that's what is required, then we might as well just let it error out. That's more in keeping with SC's general attitude of allowing users to shoot themselves in the feet. (I did misunderstand your comment -- I thought you were asking for prevention.)

up to you. i think that 'attitude' is less an explicit choice and more often lack of time to make things as robust as we'd like. which is understandable.

> But if we will have these stray actions hanging about forever, then we should more actively prevent it.

they will be reachable since they are not freed, so they will not be garbage collected. you could protect {} the wait and free in the cleanup function.

jamshark70 replied:
> they will be reachable since they are not freed, so they will not be garbage collected.

Oh gosh, good catch, yes, that should be protect-ed.

My question is really about the threads, though, for which there is no destructor. I would assume that a thread is either executing (can't be GC'ed) or scheduled on a clock (referenced from a queue, can't be GC'ed) or yielded a value that prevents rescheduling, in which case... there might be no references? (Unless the user is keeping one.) I suppose it must be like that; otherwise we'd be leaking memory like crazy.

brianlheim commented on SCClassLibrary/External/Ableton/SCClockMeterSync.sc:
```diff
@@ -0,0 +1,246 @@
+
+// TODO OSC message prefix with SC_
+
+SCClockMeterSync {
```
naive question -- why is there is `SC` prefix on this class name?

jamshark70 replied:
Link peers may include SC LinkClocks and clocks or timelines from other applications. This class doesn't (can't) try to sync any non-SC peers.

I didn't call it LinkMeterSync because I believe that it might work with the Utopia quark's BeaconClock as well. (BeaconClock promises to sync beats, but not barlines.)

Actually, I did test that just now -- BeaconClock needs to add `numPeers { ^addrBook.peers.size }` but after that, SCClockMeterSync works correctly in all respects. :open_mouth: 

So I guess I should remove references to Link from the messaging (because it *is* actually a more general solution).

telephon replied:
btw. why didn't you call it `ClockMeterSync`?

brianlheim replied:
> The "SC" prefix is meant to clarify that it's only for us -- SC clocks, not "clocks" in general.

that makes sense to me. thank you for taking the time to make this suitably generic! that's not fun work

> btw. why didn't you call it ClockMeterSync?

that's.. the exact question i asked.

jamshark70 replied:
Btw I'd welcome suggestions for other names for this class. For the 17 years I've been using SC, I've been consistently terrible at naming classes and methods, with no sign of any improvement. If you can think of something that reads better, I'd be happy to hear it. (Sincere request... `SCClockMeterSync` is rather clunky and I wouldn't mind changing the name.)

telephon replied:
I think that the difficulty in naming is an objective difficulty because programming exceeds the capacity of language. The choice is either to be as explicit as possible (and look horrible) or delegate to the background knowledge of the reader (and in lucky rare cases, find an elegant expression). I would go for the former in rarely/internally used names, the latter for  everyday code.

We don't usually make it explicit that only SC is concerned, only if something is specifically for something else, so I find the SC a bit redundant. Apart from this, the name seems fine. Alternatives might be `SharedMeter` / `SharedMetre` / `SharedMetrum`, `NetMetre` if you look for something shorter.

jamshark70 replied:
On further consideration, MeterSync is probably fine and leave it to documentation for the specifics. The one thing I might add to the documentation is to be clear about the fact that MeterSync assumes it's the clock's responsibility to sync beats and tempo (no TempoClock support, because a/ different machines, if beats aren't synced, you can't sync barlines at all and b/ same sclang instance, there's no reason to have two TempoClocks running identically -- just use the one).

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+				)
+			},
+			{
+				this.assert(false, "Make sure no other LinkClocks are running while testing");
```
in general, we assume that users are not running other servers or sc processes while testing. guarding against every possible case of that would just be took much work. i'd prefer not to have these checks, but maybe i'm missing something. could this be affected by people running LinkClocks on the local network on another machine?

jamshark70 replied:
> could this be affected by people running LinkClocks on the local network on another machine?

Yes, it will be. I do understand the general assumption but LinkClock strikes me as a special case, as it's designed to seek out peers actively on the LAN.

Will go over these comments in the next couple of days.

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+				};
+				cond.signal;
+			}.fork(clock2);
+			20.do {
```
since you use it in 4 places, probably good to give `20` a meaningful name

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+	// this test uses "magic number" wait statements
+	// because it takes a little time for a new LinkClock to stabilize
+	// and we have no signal to know when it's ready
+	test_LinkClock_sync_meter_aligns_barlines {
```
this test causes extra printing when run: "syncing meter to 3, base = 2.0". can you find a way to avoid that please?

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+				};
+				this.assertEquals(
+					list.count { |bool| bool },
+					list.size,
```
it would be simpler to keep an integer variable that you increment for every successful trial, and then assert that its value is equal to `numTrials`.

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+			outerCond.unhang;
+		});
+		outerCond.hang;
+		1.0.wait;
```
not sure i understand why you wait here

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+
+	test_LinkClock_sync_meter_propagates_meter_changes {
+		var clock1 = LinkClock.new.enableMeterSync,
+		clock2, resp, cond = Condition.new, result;
```
you never use `result`

brianlheim commented on testsuite/classlibrary/TestLinkClock.sc:
```diff
@@ -202,4 +202,140 @@ TestLinkClock : UnitTest {
+			this.assert(false, "Make sure no other LinkClocks are running while testing");
+		};
+		1.0.wait;
+		clock1.stop; clock2.stop;
```
please put separate statements on separate lines. also, same question about the wait here.

brianlheim commented:
forgot to mention in my review notes -- everything else in the C++ and sclang code looks good to me. thanks for your patience and hard work on this, it's a huge feature! @telephon maybe you want to give it a look too? i'm not an expert when it comes to clock-related things.

telephon commented on HelpSource/Classes/LinkClock.schelp:
```diff
@@ -125,6 +145,34 @@ peers join or leave the session, the LinkClock broadcasts a
+Remove all barline-sync objects.
+
+method:: syncMeter
+returns:: A Boolean, code::true:: if barline sync is active, code::false:: if not. (If you create the barline sync object independently, this answer is likely to be incorrect.)
```
what about `isRunning`?

brianlheim replied:
can't a LinkClock be running without actively syncing meter?

jamshark70 replied:
> can't a LinkClock be running without actively syncing meter?

Yes. And I'm pretty sure LinkClock already inherits isRunning from its parent TempoClock.

`isSyncingMeter` is ok with me.

jamshark70 commented:
> i think MeterSync.sc now needs to be moved to a folder other than External/Ableton. if it's meant to be a generic solution, it should be installed even if someone builds without Ableton support.

OK, into SCClassLibrary/Core (alongside Clock.sc)? But it doesn't apply to any classes in Core... Or Control? DefaultLibrary? Or maybe delete it from here and make it a quark.

Not sure what is best.

jamshark70 commented:
Except for moving MeterSync.sc, I think I got them all...?

brianlheim commented:
> OK, into SCClassLibrary/Core (alongside Clock.sc)? But it doesn't apply to any classes in Core... Or Control? DefaultLibrary? Or maybe delete it from here and make it a quark.
> Not sure what is best.

i don't have a strong opinion, but `SCClassLibrary/Common/Control` seems like a decent place for it. `Core` seems to be for more, well, core functionality, and `DefaultLibrary` doesn't seem to have much in it at all. `Control` has classes like node allocators and the server watcher, which seem like similar classes to me in that they mediate between sclang and some other process.

jamshark70 commented:
OK, I moved the file into Common/Control.

Also, after editing the unit tests yesterday, I tried them again today: a/ after deliberately breaking MeterSync (so that it doesn't change the clock's state), unit tests fail; b/ after restoring the proper code, unit tests pass.

To speed up unit tests, I reduced the timeout to look for Link peers to 0.25 instead of 1.0. The first LinkClock to set up MeterSync *must* wait this long. I haven't tested how long it takes Link to find peers on a second machine. I expect it will be faster than 250 ms, but I might be pushing the lower limit here. Should I just raise it to 0.5? That would add 0.25 sec to the two unit tests related to meter sync but would be a little safer for users.

With that said... I'm afraid I just discovered a race condition. 

1. Machine A does `LinkClock.new.enableMeterSync` and sets a different meter or barline.

2. Machines B and C do `LinkClock.new.enableMeterSync` within 0.25 seconds of each other.
   a. Machine B reads A and C, with different meters/barlines --> failed.
   b. Machine C reads A and B, with different meters/barlines --> failed.

To be correct, once one machine does `enableMeterSync`, other machines should block until that's finished, and each machine's request should listen only to the ones that have already finished meter sync.

Before I try to hack something up, let me ask the computer-science people in the room if there are any known approaches for taking a lock across the network? Or do we just document this as something to be careful about?

brianlheim commented:
w/r/t to the other two issues you mentioned, let's tackle them in different tickets/PRs. this one is in a good state already.

joshpar commented:
Across the network, @jamshark70 - I can't think of anything like that which would be standard or within the realm of our work / codebase. I think a warning is fine. I agree with @brianlheim - let's get this in for now and consider these cases more in another PR if it looks like it is needed. Seems like an issue of looking for trouble before we would know there IS trouble? IS there a race condition that you find? If so - I imagine a more useful model would be master / slave relationships (like in MIDI)... I can't think of anything that would create that kind of checking in networking code.

jamshark70 commented:
> that is to say, i'd really like to not do another round of review in this thread.

Heartily agreed :laughing:

> Seems like an issue of looking for trouble before we would know there IS trouble? IS there a race condition that you find?

I can reproduce it like this:

```
// A. sclang running in one terminal window
l = LinkClock.new.latency_(s.latency).setMeterAtBeat(3, 1).enableMeterSync;

// B. In SCIDE (launch IDE first so that this one has port 57120)
(
o = OSCFunc({
	{
		l = LinkClock.new.latency_(s.latency).enableMeterSync;
	}.defer(0.15);
}, '/sc_startedLink').oneShot;
)

// C. A second command line sclang (other terminal window)
l = LinkClock.new.latency_(s.latency).enableMeterSync; NetAddr("127.0.0.1", 57120).sendMsg('/sc_startedLink');
```

B and C both throw "LinkClock peers disagree on beatsPerBar; cannot sync barlines."

> If so - I imagine a more useful model would be master / slave relationships (like in MIDI)... I can't think of anything that would create that kind of checking in networking code.

Link is deliberately not designed after this model. One benefit of a "network of peers" is that any machine can crash, restart, and rejoin the network without disrupting the others -- whereas, in a master/slave model, if the master crashes, then you're toast.

One thing I can imagine is to keep a registry of the order in which MeterSync object IDs became active, and then sync to the oldest one (instead of the current logic, which looks for consensus).

joshpar commented:
> Link is deliberately not designed after this model. One benefit of a "network of peers" is that any machine can crash, restart, and rejoin the network without disrupting the others -- whereas, in a master/slave model, if the master crashes, then you're toast.
> 
> One thing I can imagine is to keep a registry of the order in which MeterSync object IDs became active, and then sync to the oldest one (instead of the current logic, which looks for consensus)

Makes sense - but does it have an API for you to plug in to that manages that? If it doesn't, I'm not sure that is SC's job - seems like a feature request for the Link devs (in my opinion) that, once it is architected, SC could create an implementation for.

jamshark70 commented:
> Makes sense - but does it have an API for you to plug in to that manages that? If it doesn't, I'm not sure that is SC's job - seems like a feature request for the Link devs (in my opinion) that, once it is architected, SC could create an implementation for.

I discussed it with them. I'm pretty sure they won't do that.

https://github.com/Ableton/link/issues/66#issuecomment-490520655

> Unfortunately Link can not provide continuous beat times when changing the quantum...
> 
> Here is what Link does: The quantum is used to align the downbeat to other peers that use the same quantum. As I said in the other thread, the quantum is only used when calculating the local state. It is never shared on the network and quantum changes are not orchestrated with other peers. It is used to determine a grid that tells us how to align the timelines of the individual peers, but there is no "master" timeline for all peers. And the point the individual timelines are aligned to, is opaque from the API.
> 
> So in the examples above the new timeline for the new quantum is gap-less in case the change coincidentally happens at a beat time, that is a downbeat for both, the old and new quantum. But there is no way to assert that as a user of the API.
> 
> Here is how we deal with a related scenario in Live: We set the quantum when starting playback. In case the time signature in the arrangement timeline changes, we still query the beat time using the initial quantum and just iterate beat time from there - accepting that the phase of the new time signature might not match phase with peers using the same time signature.

... which is what this PR does: when SC changes beatsPerBar, we (now) do *not* set the quantum. Then it becomes our responsibility to manage baseBarBeat (which MeterSync does).

I'm not sure I completely agree with their reasoning (they could define the Link timeline as beat, phase, quantum *and* baseBarBeat, instead of only the first three), but from their answer, I don't read anything like "we are thinking about how to improve this." Rather, it looks like "this is how it is, and you have to work with it."
